newlib/libc/stdio/nano-vfprintf_i.c

```diff
@@ -104,6 +105,20 @@ _printf_common (struct _reent *data,
 error:
   return -1;
 }
+
+// Not defined in the ESP8266 Arduino
+static const char *memchr_P(const char *src, int c, size_t length)
```

Comment:
The dummy definition (or an alias to memchr) can be put in the same file where memchr is defined.
The real definition for xtensa machine can live in xtensa directory. It can also possibly use some knowledge about progmem, i.e. that it is possible to read a 32-bit word once and then check 4 of its bytes, instead of calling pgm_read_byte which will do 4 32-bit reads.

Edit: however, i have not looked at the code generated by GCC for this. Perhaps it optimizes out unneeded loads...

Reply:
I don't think GCC looks into the assembly itself as the __asm blocks have to manually define what registers they trounce/etc.  Optimization's fine and not too complicated, but for starters I wanted something I could reasonably prove was correct. :)  This is the actual memchr routine built in the lib (modulo s/*x/pgm_read_byte(&x)/g).  The data should come from the I$ anyway given the tight look, so it's not going to cause 4 SPI reads, at least.

Reply:
Right, it's more of a thing to look into later. I am okay with merging the obviously correct version.